### PR TITLE
Refactor Tuple annotation.

### LINKED ISSUE #45: TarantoolTemplate.call does not support result mapping when Lua function returns a 'map' instead of a 'tuple'.
The fix for this issue may also deal with #32 and #28. 
Please check cases there.

Scenario:
1. Setup a cluster with roles not using CRUD library.
2. Create lua method:
```lua
function test()
 return { 
    testString = "hello world",
    testInt = 123,
}
end
```
3. Invoke TarantoolTemplate.call() to get result

```java

public Class TestEntity {
  String testString ;
  Integer testInt;
}
// ...
TestEntity t = tarantoolTemplate.call("test");

```

### PR DESCRIPTION
**Pull request by wey1and**
Closes #45
Added the ability to accept objects from stored functions of the Tarantool instance. 
Changed logic for Tuple annotation.

### CODE REVIEW

File: src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.springframework.data.tarantool.repository;
+
+import org.springframework.data.tarantool.entities.SampleUser;
+
+
+/**
+ * @author Oleg Kuznetsov
+ */
+public interface SampleUserRepository extends TarantoolRepository<SampleUser, String> {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java:9]: Implicit methods probably must raise a sane exception here.
Add TCs for error cases

File: src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java [hunk 2]
```diff
@@ -137,4 +145,65 @@ public void testBatchSave() {
         List<Book> savedBooks = bookRepository.batchSave(Arrays.asList(book1, book2));
         assertTrue(savedBooks.size() > 0);
     }
+
+    @Test
+    public void testInvocationInternalFunctionWithoutTupleAnnotationShouldReturnResponseAsObject() {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java:150]: InvocationInternalFunction -> CustomQueryMethod

File: src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java [hunk 3]
```diff
@@ -137,4 +145,65 @@ public void testBatchSave() {
...
+                .name("Vasya")
+                .lastName("Vasiliev")
+                .build();
+
+        SampleUser sampleUser = sampleUserRepository.findByName("test");
+        assertThat(sampleUser).isEqualTo(expected);
+    }
+
+    @Test
+    public void testSaveEntityWithoutTupleAnnotationShouldSaveAndReturnResponse() {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java:161]: I suggest to write comments in style of Given, When, Then.
Here it will be like:
// When Tuple is specified on Repository
// Then implicit save returns a valid object

OR rename method to 

testImplicitSaveWhenTupleIsOnRepository

File: src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java [hunk 4]
```diff
@@ -137,4 +145,65 @@ public void testBatchSave() {
...
+        BookWithoutTupleAnnotation book = BookWithoutTupleAnnotation.builder()
+                .id(4).name("Tales").uniqueKey("udf65").author("Grimm Brothers").year(1569).build();
+        BookWithoutTupleAnnotation newBook = bookRepositoryWithTuple.save(book);
+        assertThat(newBook).isEqualTo(book);
+    }
+
+    @Test
+    public void testFindByIdForEntityWithoutTupleAnnotationShouldReturnFoundedEntity() {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java:169]: testImplicitFindWhenTupleIsOnRepository

File: src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java [hunk 5]
```diff
@@ -137,4 +145,65 @@ public void testBatchSave() {
...
+                .id(6).name("The Wizard of the Emerald Town")
+                .uniqueKey("udf67").author("Leonid Volkov").year(1978).build();
+        BookWithoutTupleAnnotation newBook = bookRepositoryWithTuple.save(book);
+        bookRepositoryWithTuple.delete(newBook);
+        assertThat(bookRepositoryWithTuple.existsById(6)).isFalse();
+    }
+
+    @Test
+    @DisplayName("findById for entity without tuple annotation and with tuple on repository method " +
+            "should return founded entity")
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/support/RepositoryIntegrationTest.java:189]: founded -> found

File: src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithTuple.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.springframework.data.tarantool.core.mapping.Tuple;
+import org.springframework.data.tarantool.entities.BookWithoutTupleAnnotation;
+
+
+/**
+ * @author Oleg Kuznetsov
+ */
+@Tuple("test_space")
+public interface BookRepositoryWithTuple extends TarantoolRepository<BookWithoutTupleAnnotation, Integer> {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithTuple.java:11]: Please add custom method without @Tuple to this repository.
TC must call the method and check that it returns an object.

**Review by vrogach2020 — CHANGES_REQUESTED** [review]

File: src/test/java/org/springframework/data/tarantool/entities/Book.java [hunk 7]
```diff
@@ -15,7 +14,6 @@
  */
 @Data
 @Builder
-@EqualsAndHashCode
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/entities/Book.java:18]: Why is this removed?

File: src/test/java/org/springframework/data/tarantool/entities/SampleUser.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package org.springframework.data.tarantool.entities;
+
+import lombok.Builder;
+import lombok.Data;
+import org.springframework.data.tarantool.core.mapping.Tuple;
+
+
+/**
+ * Returned object from internal tarantool function
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/entities/SampleUser.java:9]: Not a very comprehensible description. I think it may be omitted here

File: src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithoutTuple.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * @author Oleg Kuznetsov
+ */
+public interface BookRepositoryWithoutTuple extends TarantoolRepository<Book, Integer> {
+
+    @Override
+    @Tuple("test_space")
+    Optional<Book> findById(Integer id);
+
+    @Override
+    @Tuple("test_space")
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithoutTuple.java:18]: Add a method without `@Tuple` annotation and a test for it

File: src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java (hunk 1 above)
Comment (akudiyar) [reply src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java:11]: What exception?

File: src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        tarantoolContainer.executeScript("test_setup.lua").get();
+    }
+
+    @AfterAll
+    public static void tearDown() throws Exception {
+        tarantoolContainer.executeScript("test_teardown.lua").get();
+    }
+
+    @Test
+    @DisplayName("Invocation internal function without tuple annotation should return response as object")
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java:50]: I don't get that display name, is it rrally necessary? I think it's better to name the test methods according to TCs.

Also, some comments n ncode will be helpful (see @vrogach2020 comments):
"When a method without Tuple annotation is invoked"
"Then the returned result is parsed as an object"

File: src/main/java/org/springframework/data/tarantool/core/TarantoolTemplate.java [hunk 11]
```diff
@@ -218,92 +215,120 @@ private TupleOperations setNonNullFieldsFromTuple(TarantoolTuple tuple) {
...
     @Override
-    public <T> List<T> callForList(String functionName, List<?> parameters, Class<T> entityClass) {
+    public <T> List<T> callForTupleList(String functionName, List<?> parameters, Class<T> entityClass) {
         Assert.hasText(functionName, "Function name must not be null or empty!");
         Assert.notNull(parameters, "Parameters must not be null!");
         Assert.notNull(entityClass, "Entity class must not be null!");
 
         return executeSync(getResultSupplier(
-                functionName, parameters, tarantoolClient.getConfig().getMessagePackMapper(), entityClass)
+                functionName, parameters, getMessagePackMapper(), entityClass)
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolTemplate.java:280]: Probably, `getResultSupplier` needs refactoring and merging with the new method cases. Take a look at it

**Review by akudiyar — CHANGES_REQUESTED** [review]
I'd also suggest adding a description of how the `@Tuple` annotation works to the README, and some examples of using the `call*` methods in conjunction with the Lua stored function examples.

File: src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java (hunk 1 above)
Comment (vrogach2020) [reply src/test/java/org/springframework/data/tarantool/repository/SampleUserRepository.java:11]: @akudiyar it seems the comment is outdated. It was for the case when @Tuple is not present anywhere, so implicit methods will not work because they do not the actual space to take data from.

File: src/test/resources/cartridge/app/roles/api_router.lua [hunk 12]
```diff
@@ -45,116 +49,132 @@ end
...
 function returning_error()
-	return nil, AssertionError:new('some error')
+    return nil, AssertionError:new('some error')
+end
+
+function returning_sample_user_object()
+    return { name = "Vasya", lastName = "Vasiliev" }
+end
+
+function get_all_users()
```
Comment (vrogach2020) [inline src/test/resources/cartridge/app/roles/api_router.lua:154]: I think the better name is get_predefined_users. 
It seems that all_users must select all data from repo.

File: src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                .lastName("Vasiliev")
+                .build();
+
+        SampleUser secondExpected = SampleUser.builder()
+                .name("Test")
+                .lastName("Testov")
+                .build();
+
+        //when
+        List<SampleUser> sampleUsers = sampleUserRepository.getAllUsers();
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java:78]: Please write a comment that returned data is ordered and taken from predefined lua function (not from space)

File: src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithTuple.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.springframework.data.tarantool.core.mapping.Tuple;
+import org.springframework.data.tarantool.entities.Book;
+
+
+/**
+ * @author Oleg Kuznetsov
+ */
+@Tuple("test_space")
+public interface BookRepositoryWithTuple extends TarantoolRepository<Book, Integer> {
```
Comment (vrogach2020) [inline src/test/java/org/springframework/data/tarantool/repository/BookRepositoryWithTuple.java:11]: Class BookRepository has a @Tuple annotation but without parameters. Maybe rename them  to BookRepositoryWithTupleWithSpaceParam and BookRepositoryWithTupleWithDefaults ? 
Also then we need testcases for a repository without @Tuple..

**Review by vrogach2020 — CHANGES_REQUESTED** [review]

File: src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java (hunk 13 above)
Comment (wey1and) [reply src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java:78]: I will change name of this function to get_predefined_user, that should be enough?

File: src/main/java/org/springframework/data/tarantool/core/query/TarantoolRepositoryQueryObjectExecutor.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Override
+    public Object execute(Object[] parameters) {
+        final Class<?> domainClass = queryMethod.getResultProcessor().getReturnedType().getDomainType();
+
+        if (queryMethod.isCollectionQuery()) {
+            return operations.callForObjectList(queryMethod.getQueryFunctionName(), Arrays.asList(parameters), domainClass);
+        }
+
+        return operations.callForObject(queryMethod.getQueryFunctionName(), Arrays.asList(parameters), domainClass);
```
[the remark this answers is not in the document]
Comment (wey1and) [reply src/main/java/org/springframework/data/tarantool/core/query/TarantoolRepositoryQueryObjectExecutor.java:26]: I agree with @vrogach2020 in this point
See below his comment

File: src/test/java/org/springframework/data/tarantool/core/TarantoolTemplateTest.java [hunk 16]
```diff
@@ -237,38 +256,44 @@ void testNonEntityAsReturnType_shouldHandleError() {
...
     @Test
-    public void testLoadWithCustomType() {
+    public void test_callForTuple_shouldInsertBookWithCustomType() {
+        //given
         int bookId = 12453;
         String issueDate = LocalDate.now().toString();
-        tarantoolOperations.call("insert_book_with_custom_type", Arrays.asList(bookId, issueDate), Book.class);
+
+        //when
+        tarantoolOperations.callForTuple("insert_book_with_custom_type", Arrays.asList(bookId, issueDate), "test_space", Book.class);
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/core/TarantoolTemplateTest.java:292]: Add tests for methods without specifying spaceName, as it worked before

File: src/test/java/org/springframework/data/tarantool/entities/SampleUser.java [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Builder;
+import lombok.Data;
+import org.springframework.data.tarantool.core.mapping.Tuple;
+
+
+/**
+ * @author Oleg Kuznetsov
+ */
+@Data
+@Tuple("test_space")
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/entities/SampleUser.java:12]: is it the correct space? Not "customers"?

File: src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+    public void test_save_shouldSaveBook_ifRepositoryMethodWithTupleAnnotation() {
+        //when
+        Book newBook = bookRepositoryWithoutTuple.save(this.book);
+        //then
+        assertThat(newBook).isEqualTo(book);
+    }
+
+    @Test
+    public void test_findBookByAuthor_shouldThrowException_ifTupleAnnotationOnRepository() {
+        assertThrows(InvalidDataAccessResourceUsageException.class,
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/repository/support/IntegrationTestWithoutTuples.java:140]: I think a comment is necessary here, the case is not clear

**Review by akudiyar — CHANGES_REQUESTED** [review]

File: README.md [hunk 19]
```diff
@@ -217,13 +219,88 @@ The corresponding function in on Tarantool Cartridge router may look like (uses
...
+public interface BookRepository extends TarantoolRepository<Book, Integer> {
+    
+    // it overrides schema name to test_space for current method
+    @Tuple("test_space") 
+    @Query(function = "do_something")
+    Optional<Book> doSomething(Book book);
+}
+```
+
+By default, schema name is book_entity because schema name not specified
```
Comment (akudiyar) [inline README.md:267]:
```suggestion
Here the space name is not specified neither in a `@Tuple` annotation on the repository interface, nr on the entity class, so it will be "book_entity" according to the entity class name, turned into "lower snake case".
```

File: README.md [hunk 20]
```diff
@@ -217,13 +219,88 @@ The corresponding function in on Tarantool Cartridge router may look like (uses
...
+    
+    // it overrides schema name to test_space for current method
+    @Tuple("test_space") 
+    @Query(function = "do_something")
+    Optional<Book> doSomething(Book book);
+}
+```
+
+By default, schema name is book_entity because schema name not specified 
+on the entity class and on the repository. In this case schema name being
```
Comment (akudiyar) [inline README.md:268]:
```suggestion
```

File: README.md [hunk 21]
```diff
@@ -217,13 +219,88 @@ The corresponding function in on Tarantool Cartridge router may look like (uses
...
+    // it overrides schema name to test_space for current method
+    @Tuple("test_space") 
+    @Query(function = "do_something")
+    Optional<Book> doSomething(Book book);
+}
+```
+
+By default, schema name is book_entity because schema name not specified 
+on the entity class and on the repository. In this case schema name being 
+name of entity class in snake_case, but for "doSomething" method space name
```
Comment (akudiyar) [inline README.md:269]:
```suggestion
But for the "doSomething" method the space name will be overridden to "test_space".
```

File: README.md [hunk 22]
```diff
@@ -217,13 +219,88 @@ The corresponding function in on Tarantool Cartridge router may look like (uses
...
+    @Tuple("test_space") 
+    @Query(function = "do_something")
+    Optional<Book> doSomething(Book book);
+}
+```
+
+By default, schema name is book_entity because schema name not specified 
+on the entity class and on the repository. In this case schema name being 
+name of entity class in snake_case, but for "doSomething" method space name 
+is a test_space.
```
Comment (akudiyar) [inline README.md:270]:
```suggestion
```

File: README.md [hunk 23]
```diff
@@ -217,13 +219,88 @@ The corresponding function in on Tarantool Cartridge router may look like (uses
...
+@Tuple
+public class BookEntity {
+  @Id
+  private Integer id;
+  private String name;
+}
+
+public interface BookRepository extends TarantoolRepository<BookEntity, Integer> {
+    
+    // it overrides schema name to test_space for current method
```
Comment (akudiyar) [inline README.md:281]:
```suggestion
    // it overrides the space name to "test_space" for the current method
```

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolOperations.java [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+     * @param <T>        target entity type
+     * @param <ID>       target entity index type
+     * @param id         Target entity ID
+     * @param entityType Desired type of the result object
+     * @return Removed entity value
+     */
+    @Nullable
+    <T, ID> T removeById(ID id, Class<T> entityType);
+
+    TarantoolMappingContext getMappingContext();
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolOperations.java:152]: Javadoc is missing

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 25]
```diff
@@ -0,0 +1,480 @@
...
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.data.tarantool.core.convert.TarantoolConverter;
+import org.springframework.data.tarantool.core.mapping.TarantoolMappingContext;
+import org.springframework.data.tarantool.core.mapping.TarantoolPersistentEntity;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+import java.util.*;
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:29]: Expand asterisk import

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 26]
```diff
@@ -0,0 +1,480 @@
...
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.springframework.data.tarantool.core.TarantoolTemplateUtils.getIndexPartsFromCompositeIdValue;
+import static org.springframework.data.tarantool.core.TarantoolTemplateUtils.idQueryFromTuple;
+
+/**
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:41]: Add description, what methods this class contains

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 27]
```diff
@@ -0,0 +1,480 @@
...
+                tarantoolClient.space(entity.getSpaceName()).select(query)
+        );
+        return result.stream().map(t -> mapToEntity(t, entityClass)).collect(Collectors.toList());
+    }
+
+    @Override
+    public <T, ID> T findById(ID id, Class<T> entityClass) {
+        Assert.notNull(id, "Id must not be null!");
+        Assert.notNull(entityClass, "Entity class must not be null!");
+
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:99]: Extra newline

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 28]
```diff
@@ -0,0 +1,480 @@
...
+                if (wrapped != null) {
+                    throw wrapped;
+                }
+            }
+            throw new DataRetrievalFailureException(e.getMessage(), e.getCause());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:393]: Extra newline

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 29]
```diff
@@ -0,0 +1,480 @@
...
+    protected <T> T mapToEntity(@Nullable Object tuple, Class<T> entityClass) {
+        return getConverter().read(entityClass, tuple);
+    }
+
+    protected Conditions createIndexEqualsConditionFromParts(List<?> indexPartValues) {
+        List<?> indexPartValuesConverted = mapParameters(indexPartValues);
+        return Conditions.indexEquals(TarantoolIndexQuery.PRIMARY, indexPartValuesConverted);
+    }
+
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:462]: Extra newline

File: src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java [hunk 30]
```diff
@@ -0,0 +1,480 @@
...
+
+    protected List<?> mapParameters(List<?> parameters) {
+        List<Object> mappedParameters = new LinkedList<>();
+        getConverter().write(parameters, mappedParameters);
+        return mappedParameters;
+    }
+
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/BaseTarantoolTemplate.java:469]: Extra newline

File: src/main/java/org/springframework/data/tarantool/core/TarantoolOperations.java [hunk 31]
```diff
@@ -1,309 +1,316 @@
...
 import io.tarantool.driver.mappers.ValueConverter;
 import org.msgpack.value.Value;
-import org.springframework.data.tarantool.core.convert.TarantoolConverter;
 import org.springframework.lang.Nullable;
 
 import java.util.List;
 
 /**
- * Interface that specifies a basic set of Tarantool operations. Implemented by {@link TarantoolTemplate}.
+ * Interface that specifies a set extensions of Tarantool operations. Implemented by {@link TarantoolTemplate}.
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolOperations.java:10]:
```suggestion
 * Interface that specifies a set of Tarantool call operations extensions. Implemented by {@link TarantoolTemplate}.
```

File: src/main/java/org/springframework/data/tarantool/core/TarantoolOperations.java [hunk 32]
```diff
@@ -1,309 +1,316 @@
...
  *
  * @author Alexey Kuzin
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolOperations.java:12]: Add `@author` here

File: src/main/java/org/springframework/data/tarantool/core/TarantoolTemplate.java [hunk 33]
```diff
@@ -1,505 +1,143 @@
...
-    private static final int MAX_WORKERS = 4;
-
-    private final TarantoolClient<TarantoolTuple, TarantoolResult<TarantoolTuple>> tarantoolClient;
-    private final TarantoolMappingContext mappingContext;
-    private final TarantoolConverter converter;
-    private final TarantoolExceptionTranslator exceptionTranslator;
-    private final ForkJoinPool queryExecutors;
-    private final MessagePackMapper mapper;
+/**
+ * @author Alexey Kuzin
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolTemplate.java:17]: Add a description about what methods this class contains

File: src/main/java/org/springframework/data/tarantool/core/TarantoolTemplateUtils.java [hunk 34]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.data.tarantool.core.mapping.TarantoolPersistentProperty;
+import org.springframework.util.Assert;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+
+/**
+ * @author Oleg Kuznetsov
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolTemplateUtils.java:16]: Add a description about what methods this class contains

File: src/main/java/org/springframework/data/tarantool/core/TarantoolTemplateUtils.java [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+                throw new MappingException("No ID property specified in persistent entity " + entity.getType());
+            }
+            Object fieldValue = tuple.getObject(idProperty.getFieldName(), idProperty.getType())
+                    .orElseThrow(() -> new MappingException("ID property value is null"));
+            idValue = Collections.singletonList(fieldValue);
+        }
+        return Conditions.indexEquals(0, idValue);
+    }
+
+    static List<?> getIndexPartsFromCompositeIdValue(Object idValue, TarantoolPersistentEntity<?> entity) {
```
Comment (akudiyar) [inline src/main/java/org/springframework/data/tarantool/core/TarantoolTemplateUtils.java:34]: Add a Javadoc for this method

File: src/test/java/org/springframework/data/tarantool/TestUtils.java [hunk 36]
```diff
@@ -0,0 +1,37 @@
+package org.springframework.data.tarantool;
+
+import lombok.SneakyThrows;
+
+import java.util.concurrent.Callable;
+
+public class TestUtils {
```
Comment (akudiyar) [inline src/test/java/org/springframework/data/tarantool/TestUtils.java:7]: Where is it used?

FYI there is a retry mechanism in the driver, you may find it helpful.

**Review by akudiyar — CHANGES_REQUESTED** [review]

**Review by vrogach2020 — APPROVED** [review]

**Review by akudiyar — APPROVED** [review]
